File: src/Configuration.vb
Imports Inventor

Public Class Configuration

    Public ConfigName As String
    Public UserName As String '
    Public DefaultVBAProjectFileFullFilename As String '
    Public TemplatesPath As String '
    Public DesignDataPath As String '
    Public PresetsPath As String '
    Public SymbolLibraryPath As String '
    Public SheetMetalPunchesRootPath As String '
    Public iFeatureOptionsRootPath As String '
    Public CCAccess As String '
    Public CCLibraryPath As String '
    Public CCCustomFamilyAsStandard As Boolean '
    Public CCRefreshOutOfDateStandardParts As Boolean '
    Public SectionAllParts As Boolean '
    Public DefaultDrawingFileType As String '
    Public ExternalRuleDirectories() As String 'TODO!

    Private Application As Inventor.Application

    Public Sub Apply(ByRef app As Inventor.Application)
        Application = app

        SetStringOption(UserName, Application.GeneralOptions.UserName)
        SetStringOption(DefaultVBAProjectFileFullFilename, Application.FileOptions.DefaultVBAProjectFileFullFilename)
        SetStringOption(TemplatesPath, Application.FileOptions.TemplatesPath)
        SetStringOption(DesignDataPath, Application.FileOptions.DesignDataPath)
        SetStringOption(PresetsPath, Application.FileOptions.PresetsPath)
        SetStringOption(SymbolLibraryPath, Application.FileOptions.SymbolLibraryPath)
        SetStringOption(SheetMetalPunchesRootPath, Application.iFeatureOptions.SheetMetalPunchesRootPath)
        SetStringOption(iFeatureOptionsRootPath, Application.iFeatureOptions.RootPath)

        SetCCAccessOption(CCAccess, CCLibraryPath)
        SetBoolOption(CCCustomFamilyAsStandard, Application.ContentCenterOptions.CustomFamilyAsStandard)
        SetBoolOption(CCRefreshOutOfDateStandardParts, Application.ContentCenterOptions.RefreshOutOfDateStandardParts)

        SetBoolOption(SectionAllParts, Application.AssemblyOptions.SectionAllParts)

        SetDefaultDrawingOption(DefaultDrawingFileType, Application.DrawingOptions.DefaultDrawingFileType)

    End Sub


    Private Sub SetStringOption(ByVal prop As String, ByRef appOption As Object)

        If String.IsNullOrEmpty(prop) Then
            Return
        End If

        Try
            appOption = prop
        Catch ex As Exception

        End Try

    End Sub


    Private Sub SetCCAccessOption(ByVal access As String, ByRef path As String)

        If String.IsNullOrEmpty(access) Then Return


        Select Case access.ToLower()
            Case "desktop"
                If String.IsNullOrEmpty(path) Then Return

                Application.ContentCenterOptions.SetAccessOption(Inventor.ContentCenterAccessOptionEnum.kInventorDesktopAccess, path)

            Case "vault"
                Application.ContentCenterOptions.SetAccessOption(Inventor.ContentCenterAccessOptionEnum.kVaultOrProductstreamServerAccess)

        End Select

    End Sub

    Private Sub SetBoolOption(ByVal prop As Boolean, ByRef appOption As Object)

        Try
            appOption = prop
        Catch ex As Exception

        End Try

    End Sub

    Private Sub SetDefaultDrawingOption(prop As String, appOption As Object)

        If String.IsNullOrEmpty(prop) Then Return

        Select Case prop.ToLower()
            Case "dwg"
                appOption = 69633

            Case "idw"
                appOption = 69634

        End Select

    End Sub


End Class

File: src/ConfigLoader/Configuration.cs
using System;
using Inventor;
//using Autodesk.iLogic.Interfaces;
using System.Linq;
using System.Collections.Generic;

namespace ConfigLoader
{
    public class Configuration
    {
        public string ConfigName;
        public string UserName;
        public string DefaultVBAProjectFileFullFilename;
        public string TemplatesPath;
        public string DesignDataPath;
        public string PresetsPath;
        public string SymbolLibraryPath;
        public string SheetMetalPunchesRootPath;
        public string iFeatureOptionsRootPath;
        public string CCAccess;
        public string CCLibraryPath;
        public bool CCCustomFamilyAsStandard;
        public bool CCRefreshOutOfDateStandardParts;
        public bool SectionAllParts;
        public string DefaultDrawingFileType;
        public string[] ExternalRuleDirectories;
        public bool CleanExternalRuleDirectories;
        private Application app;
        private dynamic iLogicAuto;

        public void LoadConfigurationIntoInventor(Application application)
        {
            app = application;
            iLogicAuto = GetiLogicAddIn(app);


            SetStringOption(UserName, s => app.GeneralOptions.UserName = s);
            SetStringOption(DefaultVBAProjectFileFullFilename, s => app.FileOptions.DefaultVBAProjectFileFullFilename = s);
            SetStringOption(TemplatesPath, s => app.FileOptions.TemplatesPath = s);
            SetStringOption(DesignDataPath, s => app.FileOptions.DesignDataPath = s);
            SetStringOption(PresetsPath, s => app.FileOptions.PresetsPath = s);
            SetStringOption(SymbolLibraryPath, s => app.FileOptions.SymbolLibraryPath = s);
            SetStringOption(SheetMetalPunchesRootPath, s => app.iFeatureOptions.SheetMetalPunchesRootPath = s);
            SetStringOption(iFeatureOptionsRootPath, s => app.iFeatureOptions.RootPath = s);

            SetCCAccessOption(CCAccess, CCLibraryPath);
            SetBoolOption(CCCustomFamilyAsStandard, b => app.ContentCenterOptions.CustomFamilyAsStandard = b);
            SetBoolOption(CCRefreshOutOfDateStandardParts, b => app.ContentCenterOptions.RefreshOutOfDateStandardParts = b);

            SetBoolOption(SectionAllParts, b => app.AssemblyOptions.SectionAllParts = b);

            SetDefaultDrawingOption(DefaultDrawingFileType, l => app.DrawingOptions.DefaultDrawingFileType = (DefaultDrawingFileTypeEnum)l);

            EmptyExternalRuleDirectories();
            SetExternalRuleDirectories();

        }

        private void SetStringOption(string prop, Action<string> appOption)
        {
            if (String.IsNullOrEmpty(prop))
                return;

            try
            {
                appOption(prop);
            }
            catch (Exception e)
            {
                throw new SystemException("The value of the " + nameof(appOption) + " setting in the json configuration file was invalid.", e);
            }
        }

        private void SetCCAccessOption(string access, string path)
        {
            if (string.IsNullOrEmpty(access))
                return;

            try
            {
                switch (access.ToLower())
                {
                    case "desktop":
                        if (string.IsNullOrEmpty(path))
                            return;

                        app.ContentCenterOptions.SetAccessOption(Inventor.ContentCenterAccessOptionEnum.kInventorDesktopAccess, path);
                        break;

                    case "vault":
                        app.ContentCenterOptions.SetAccessOption(Inventor.ContentCenterAccessOptionEnum.kVaultOrProductstreamServerAccess);
                        break;
                }
            }
            catch (Exception e)
            {
                throw new SystemException("The value of the CCLibraryPath setting in the json configuration file was invalid.", e);
            }
        }

        private void SetBoolOption(bool prop, Action<bool> appOption)
        {
            try
            {
                appOption(prop);
            }
            catch { }
        }

        private void SetDefaultDrawingOption(string prop, Action<int> appOption)
        {
            if (string.IsNullOrEmpty(prop))
                return;

            switch (prop.ToLower())
            {
                case "dwg":
                    appOption(69633);
                    break;

                case "idw":
                    appOption(69634);
                    break;
            }
        }

        /// <summary>
        /// Set the External iLogic Rule directories
        /// </summary>
//        internal IiLogicAutomation iLogicAutomation { get; private set; }
        internal void SetExternalRuleDirectories()
        {
            if (ExternalRuleDirectories.Length == 0)
                return;

            //Add iLogic Directories
            string[] oldDirectories = iLogicAuto.FileOptions.ExternalRuleDirectories;
            List<string> newDirectories = oldDirectories.OfType<string>().ToList();

            //if it isn't already in there, add it!
            foreach (string i in ExternalRuleDirectories)
            {
                if (!newDirectories.Contains(i))
                {
                    newDirectories.Add(i);
                }
            }

            try
            {
                if (newDirectories.Count > oldDirectories.Length)
                {
                    iLogicAuto.FileOptions.ExternalRuleDirectories = newDirectories.ToArray();
                }

            }
            catch (Exception e)
            {
                throw new SystemException("An error occurred while updating the External iLogic Rule Directories", e);
            }
        }

        private void EmptyExternalRuleDirectories()
        {
            if (CleanExternalRuleDirectories)
            {
                string[] temp = { };
                iLogicAuto.FileOptions.ExternalRuleDirectories = temp;
            }
        }

        internal object GetiLogicAddIn(Inventor.Application _app)
        {
            ApplicationAddIns appAddIns = _app.ApplicationAddIns;
            ApplicationAddIn addIn = appAddIns.ItemById["{3bdd8d79-2179-4b11-8a5a-257b1c0263ac}"];

            if (addIn == null)
                throw new SystemException("The iLogic add-in could not be found by ID {3bdd8d79-2179-4b11-8a5a-257b1c0263ac}.");

            try
            {
                addIn.Activate();
                return addIn.Automation;
            }
            catch (Exception e)
            {
                throw new SystemException("The iLogic add-in could not be accessed.", e);
            }
        }


    }
}
